File: Raspi/client/cpp/driver.cpp
#include "driver.h"
#include <unistd.h>
#include <algorithm>
#include <stdio.h>
#include <string.h>
#include <arpa/inet.h>
#include <fcntl.h>

using namespace std;

driver::driver(){
    printf("----Driver ID:%d Init----\n", this);

    sock = socket(PF_INET, SOCK_DGRAM, 0);
    bzero(&serv_addr, sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons(61551);
    serv_addr.sin_addr.s_addr = inet_addr("127.0.0.1");
    fcntl(sock, F_SETFL, fcntl(sock, F_GETFL, 0) | O_NONBLOCK);
    
    keepRunning = true;
    recv_thread = thread(recv_threading, this);
}

driver::~driver(){
    keepRunning = false;
    recv_thread.join();
    printf("----Driver ID:%d End----\n", this);
}

void driver::set_code(char* code, float v){
    int value = max(min(1.0f, v), -1.0f) * 100;
    sprintf(buf, "{\"%s\":%d}", code, value);
    launch();
}

void driver::setMotor(float motor){
    set_code("sm", motor);
}

void driver::setServo(float servo){
    set_code("ss", motor);
}

void driver::setCamPitch(float pitch){
    set_code("scp", pitch);
}

void driver::setCamYaw(float yaw){
    set_code("scy", yaw);
}

void driver::setMotion(float motor, float servo){
    motor = max(min(1.0f, motor), -1.0f);
    servo = max(min(1.0f, servo), -1.0f);
    sprintf(buf, "{\"sm\":%d, \"ss\":%d}",int(motor * 100), int(servo * 100));
    launch();
}

void drvier::query_code(int code){
    sprintf(buf, "{\"rq\":%d}", code);
    launch();
}

void driver::getMode(){
    query_code(0);
}

void driver::getMotor(){
    query_code(1);
}

void driver::getServo(){
    query_code(2);
}

void getCamPitch(){
    query_code(3);
}

void getCamYaw(){
    query_code(4);
}

void driver::launchVoltageMode(){
    sprintf(buf, "{\"so\":1}");
    launch();
}

void driver::allStop(){
    sprintf(buf, "{\"so\":0}");
    launch();
}

int driver::launch(){
    if (sendto(sock, buf, strlen(buf), 0, (struct sockaddr *) &serv_addr, sizeof(struct sockaddr_in))< 0){
        printf("sendto failed.\n");
        return 1;
    }
    return 0;
}

void recv_threading(driver* d){
   char buf[256];
   struct sockaddr client_addr;
   socklen_t client_len = sizeof(client_addr);
   while(d->keepRunning){
        int recv_num = recvfrom(d->sock,buf,sizeof(buf),0,&client_addr, &client_len);
        if (recv_num > 0){
            if(recv_num < sizeof(buf)){
                buf[recv_num] = 0;
            };
            printf("recv: %s\n", buf);
        }
        usleep(1000);
   }
}

File: Raspi/client/python3/driver.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

"""
@file: driver.py
@time: 2018/8/12 0:19
"""

import threading
import socket
import json
import time
import traceback


def dash_print(txt, num=7):
    return '-' * num + txt + '-' * num


class driver:

    def __init__(self):
        self.direct_launch = True
        self.__dst = ('127.0.0.1', 61551)
        self.__conf = {}
        self.__sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.__sock.settimeout(1.0)
        self.__keepRunning = False
        self.__recv_t = threading.Thread(target=self.recv_thread)
        print(dash_print('Driver No.%d Init' % id(self)))

    def __enter__(self):
        if self.__keepRunning:
            print('Driver No.%d is running.' % id(self))
        else:
            self.open()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def __del__(self):
        self.__sock.close()
        print(dash_print('Driver No.%d Deleted' % id(self)))

    def open(self):
        print(dash_print('Driver No.%d Open' % id(self)))
        try:
            self.__keepRunning = True
            self.__recv_t.start()
        except:
            traceback.print_exc()
            print(dash_print('Driver Open Failed'))
            return None
        print(dash_print('Driver Open Done'))

    def close(self):
        self.__keepRunning = False
        self.__recv_t.join()
        print(dash_print('Driver No.%d Closed' % id(self)))

    def __launch(self):
        self.__sock.sendto(json.dumps(self.__conf).encode('utf-8'), self.__dst)
        self.__conf = {}

    def __set_code(self, code, v):
        self.__conf[code] = round(max(min(v, 1), -1) * 100)
        if self.direct_launch:
            self.__launch()

    def setMotor(self, motor):
        self.__set_code('sm', motor)

    def setServo(self, steer):
        self.__set_code('ss', steer)

    def setCamPitch(self, steer):
        self.__set_code('scp', steer)

    def setCamYaw(self, steer):
        self.__set_code('scy', steer)

    def setMode(self, md):
        if md == 'voltage':
            self.__conf['so'] = 1
        if md == 'stop':
            self.__conf['so'] = 0

    def __query_code(self, code):
        self.__conf['rq'] = code
        self.__launch()

    def getMode(self):
        self.__query_code(0)

    def getMotor(self):
        self.__query_code(1)

    def getServo(self):
        self.__query_code(2)
        
    def getCamPitch(self):
        self.__query_code(3)

    def getCamYaw(self):
        self.__query_code(4)

    def setStatus(self, **dt):
        self.__conf['uid'] = round(time.time(), 5)
        self.direct_launch = False
        if 'motor' in dt:
            self.setMotor(dt['motor'])
        if 'servo' in dt:
            self.setServo(dt['servo'])
        if 'mode' in dt:
            self.setMode(dt['mode'])
        if 'cam_pitch' in dt:
            self.setCamPitch(dt['cam_pitch'])
        if 'cam_yaw' in dt:
            self.setCamYaw(dt['cam_yaw'])
        self.__launch()
        self.direct_launch = True
    
    def heartBeat(self):
        self.setStatus()

    def parse_feedback(self, js):
        try:
            obj = json.loads(js.decode('utf-8'))
            """
            if not 'uid' in obj:
                print("Uid is needed")
                return 2
            elif self.__uid > obj['uid']:
                print("Uid %0.3f is behind %0.3f, drop %s." % (obj['uid'], self.__uid, js))
            else:
                self.__uid = float(obj['uid']
            """
            return 0
        except:
            traceback.print_exc()
            return 1

    def recv_thread(self):
        while self.__keepRunning:
            try:
                res = self.__sock.recvfrom(1024)
                print("recv: %s, from %s:%d" % (res[0], res[1][0], res[1][1]))
                if self.parse_feedback(res[0]):
                    print("Error feedback: %s" % res[0])
            except socket.timeout:
                pass
            except:
                traceback.print_exc()
                self.__keepRunning = False
        print(dash_print('Driver No.%d Recv closed' % id(self)))


if __name__ == '__main__':
    print(dash_print("Simple Raspi Car Driver"))
    print("Usage: import driver.py and using driver() to start.")
    print("-" * 10)

